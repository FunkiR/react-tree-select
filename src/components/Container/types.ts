import type {ComponentType, HTMLProps} from 'react';

export type Props = HTMLProps<HTMLDivElement>;

export type ContainerComponent = ComponentType<Props>;
