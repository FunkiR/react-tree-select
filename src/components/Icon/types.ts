import {FontAwesomeIconProps} from '@fortawesome/react-fontawesome';
import {ComponentType} from "react";

export type Props = FontAwesomeIconProps;

export type IconComponent = ComponentType<Props>;
