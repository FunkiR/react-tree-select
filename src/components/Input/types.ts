export type OnChangeEvent = {
	name: string,
	value: string
};

export type Props = {
    className?: string,
	name?: string,
    onChange: (event: OnChangeEvent) => void,
    value: string
};
