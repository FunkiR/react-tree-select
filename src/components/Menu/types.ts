import {ComponentType} from "react";
import type {InternalOnExpandEvent,Node} from '~/types';

export type Props = {
	className: string,
	isSelected: (node: Node) => boolean,
	tree: Node[],
	onExpand: (event: InternalOnExpandEvent) => void,
	onSelect: (node: Node) => void
};

export type MenuComponent = ComponentType<Props>;
