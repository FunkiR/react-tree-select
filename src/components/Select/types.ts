import {ComponentType} from "react";
import {Node, OnChangeEvent, OnExpandEvent, OnSelectEvent, Value} from '~/types';

export type Props = {
	className: string,
	name: string,
	isSelected?: (node: Node, value: Value | null) => boolean,
	onChange: (event: OnChangeEvent) => void,
	onExpand: (event: OnExpandEvent) => void,
	onSelect: (event: OnSelectEvent) => void,
	tree: Array<Node>,
	value: Value | null
};

export type SelectComponent = ComponentType<Props>;
