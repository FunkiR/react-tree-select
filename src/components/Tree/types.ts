import type {Props as MenuProps} from '~/components/Menu/types';

export type Props = Pick<MenuProps, 'isSelected' | 'onExpand' | 'onSelect' | 'tree'> & {
	searchValue: string
};
