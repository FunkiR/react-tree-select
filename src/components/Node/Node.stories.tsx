import {action} from "@storybook/addon-actions";
import {Meta, Story} from '@storybook/react/types-6-0';
import React from 'react';
import {withSelectContext} from "stories/decorators";
import {generateArrayTree} from "stories/helpers";

import Node from './Node';
import type {Props} from './types';

export default {
	title: 'Components/Node',
	component: Node,
	decorators: [withSelectContext]
} as Meta;

const Template: Story<Props> = args => <Node {...args}  />;

export const WithoutChildren = Template.bind({});

WithoutChildren.args = {
	data: {
		children: [],
		label: 'label',
		value: 'value'
	},
	onClick: action('onClick'),
	onExpand: action('onExpand'),
	hasChildren: false,
	label: 'label',
	found: false,
	selected: false
};

export const WithChildren = Template.bind({});

WithChildren.args = {
	...WithoutChildren.args,
	data: {
		children: generateArrayTree(5, 1),
		label: 'label',
		value: 'value'
	},
	children: node => node.children.map(child => (
		<Node {...WithoutChildren.args} data={child} key={child.value} label={child.label} />
	)),
	hasChildren: true
};

