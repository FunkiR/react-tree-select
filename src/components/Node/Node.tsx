import {faMinusSquare, faPlusSquare} from '@fortawesome/free-solid-svg-icons';
import cn from 'classnames';
import useSelect from "~/hooks/useSelect";
import React, {FC, memo, useCallback, useEffect, useMemo, useState} from 'react';

import type {Props} from './types';

export const Node: FC<Props> = props => {
	const {children, data, expanded: externalExpanded, found, hasChildren, label, onClick, onExpand, selected} = props;
	const [components, theme] = useSelect('components', 'theme');
	const {
		CollapsedIcon,
		ExpandedIcon,
		NodeChildrenContainer,
		NodeContainer,
		NodeExpandIconContainer,
		NodeLabel,
		NodeLabelContainer
	} = components;
	const [expanded, setExpanded] = useState(externalExpanded);

	useEffect(() => setExpanded(externalExpanded), [externalExpanded]);

	const handleClick = useCallback(() => onClick(data), [onClick, data]);

	const handleCLickExpandIcon = useCallback(() => {
		const nextExpanded = !expanded;

		setExpanded(nextExpanded);
		onExpand({
			node: data,
			expanded: nextExpanded
		});
	}, [expanded, data, setExpanded]);

	const renderChildrenNodes = () => {
		if (expanded && hasChildren) {
			return (
				<NodeChildrenContainer className={theme.NodeChildrenContainer}>
					{children(data)}
				</NodeChildrenContainer>
			);
		}

		return null;
	};

	const renderExpandIcon = () => {
		if (hasChildren) {
			return expanded
				? <ExpandedIcon icon={faMinusSquare} onClick={handleCLickExpandIcon} />
				: <CollapsedIcon icon={faPlusSquare} onClick={handleCLickExpandIcon} />;
		}

		return null;
	};

	const labelCN = useMemo(() => cn({
		[theme.NodeLabel]: true,
		[theme.NodeLabel_found]: found,
		[theme.NodeLabel_selected]: selected
	}), [found, selected]);

	return (
		<NodeContainer className={theme.NodeContainer}>
			<NodeLabelContainer className={theme.NodeLabelContainer}>
				<NodeExpandIconContainer className={theme.NodeExpandIconContainer}>
					{renderExpandIcon()}
				</NodeExpandIconContainer>
				<NodeLabel className={labelCN} onClick={handleClick} title={label}>{label}</NodeLabel>
			</NodeLabelContainer>
			{renderChildrenNodes()}
		</NodeContainer>
	);
};

export default memo(Node);
