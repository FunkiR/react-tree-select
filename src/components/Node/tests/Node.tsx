import {Node} from "components/Node/Node";
import {Props} from "components/Node/types";
import {mount, shallow} from "enzyme";
import React from "react";
import {Component} from "~/types";

describe('Node component', () => {
	const getProps = (mixin: Partial<Props> = {}) => {
		const data = {
			label: 'label',
			value: 'value',
			children: []
		};

		return {
			data,
			expanded: false,
			found: false,
			hasChildren: false,
			label: data.label,
			onClick: jest.fn(),
			onExpand: jest.fn(),
			children: jest.fn().mockReturnValue(null),
			selected: false,
			...mixin
		};
	};

	it('renders without children', () => {
		const wrapper = shallow(<Node {...getProps()} />);

		expect(wrapper).toMatchSnapshot();
	});

	describe('renders with children when "expanded" prop changes', () => {
		const props = getProps({
			data: {
				children: [
					{
						children: [],
						label: 'child1',
						value: 'child1'
					},
					{
						children: [],
						label: 'child2',
						value: 'child2'
					},
				],
				label: 'label',
				value: 'value'
			},
			hasChildren: true,
			children: node => node.children.map(child => {
				return <Node {...getProps({data: child, label: child.label})} hasChildren={false} key={child.value} />;
			})
		});

		it('expanded is false', () => {
			const wrapper = shallow(<Node {...props} expanded={false} />);

			expect(wrapper).toMatchSnapshot();
		});

		it('expanded is true', () => {
			const wrapper = shallow(<Node {...props} expanded={true} />);

			expect(wrapper).toMatchSnapshot();
		});
	});

	it('handles onClick by label and calls provided callback with node data', () => {
		const mockOnClick = jest.fn();
		const props = getProps({
			onClick: mockOnClick
		});

		const wrapper = mount(<Node {...props} />);

		wrapper.find(`div.${Component.NodeLabel}`).simulate('click');

		expect(mockOnClick).toBeCalledWith(props.data);
	});

	it('handles onExpand and calls provided callback with expanded and node data', () => {
		const mockOnExpand = jest.fn();
		const props = getProps({
			expanded: false,
			hasChildren: true,
			onExpand: mockOnExpand
		});

		const wrapper = mount(<Node {...props} />);

		wrapper.find(`div.${Component.NodeExpandIconContainer} svg`).simulate('click');

		expect(mockOnExpand).toBeCalledWith({expanded: true, node: props.data});
	});
});

