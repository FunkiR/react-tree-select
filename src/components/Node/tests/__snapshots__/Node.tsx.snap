// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node component renders correctly with children and when "expanded" prop changes expanded is false 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    >
      <Memo(Icon)
        icon={
          Object {
            "icon": Array [],
            "iconName": "plus-square",
            "prefix": "fas",
          }
        }
        onClick={[Function]}
      />
    </Memo(ForwardRef)>
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;

exports[`Node component renders correctly with children and when "expanded" prop changes expanded is true 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    >
      <Memo(Icon)
        icon={
          Object {
            "icon": Array [],
            "iconName": "minus-square",
            "prefix": "fas",
          }
        }
        onClick={[Function]}
      />
    </Memo(ForwardRef)>
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
  <Memo(ForwardRef)
    className="NodeChildrenContainer"
  >
    <Node
      data={
        Object {
          "children": Array [],
          "label": "child1",
          "value": "child1",
        }
      }
      expanded={false}
      found={false}
      hasChildren={false}
      key="child1"
      label="child1"
      onClick={[MockFunction]}
      onExpand={[MockFunction]}
      selected={false}
    >
      <Component />
    </Node>
    <Node
      data={
        Object {
          "children": Array [],
          "label": "child2",
          "value": "child2",
        }
      }
      expanded={false}
      found={false}
      hasChildren={false}
      key="child2"
      label="child2"
      onClick={[MockFunction]}
      onExpand={[MockFunction]}
      selected={false}
    >
      <Component />
    </Node>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;

exports[`Node component renders correctly without children 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    />
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;

exports[`Node component renders with children when "expanded" prop changes expanded is false 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    >
      <Memo(Icon)
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f0fe",
              "M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z",
            ],
            "iconName": "plus-square",
            "prefix": "fas",
          }
        }
        onClick={[Function]}
      />
    </Memo(ForwardRef)>
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;

exports[`Node component renders with children when "expanded" prop changes expanded is true 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    >
      <Memo(Icon)
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f146",
              "M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM92 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H92z",
            ],
            "iconName": "minus-square",
            "prefix": "fas",
          }
        }
        onClick={[Function]}
      />
    </Memo(ForwardRef)>
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
  <Memo(ForwardRef)
    className="NodeChildrenContainer"
  >
    <Node
      data={
        Object {
          "children": Array [],
          "label": "child1",
          "value": "child1",
        }
      }
      expanded={false}
      found={false}
      hasChildren={false}
      key="child1"
      label="child1"
      onClick={[MockFunction]}
      onExpand={[MockFunction]}
      selected={false}
    >
      <Component />
    </Node>
    <Node
      data={
        Object {
          "children": Array [],
          "label": "child2",
          "value": "child2",
        }
      }
      expanded={false}
      found={false}
      hasChildren={false}
      key="child2"
      label="child2"
      onClick={[MockFunction]}
      onExpand={[MockFunction]}
      selected={false}
    >
      <Component />
    </Node>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;

exports[`Node component renders without children 1`] = `
<Memo(ForwardRef)
  className="NodeContainer"
>
  <Memo(ForwardRef)
    className="NodeLabelContainer"
  >
    <Memo(ForwardRef)
      className="NodeExpandIconContainer"
    />
    <Memo(ForwardRef)
      className="NodeLabel"
      onClick={[Function]}
      title="label"
    >
      label
    </Memo(ForwardRef)>
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;
